reference/src/clientimpl/src/sa_key_provision.c

```diff
@@ -0,0 +1,364 @@
+   }
+
+   switch(ta_key_type) {
+      case WIDEVINE_OEM_PROVISIONING:
+         {
+            DEBUG(" case WIDEVINE_OEM_PROVISIONING");
+
+            if(NULL == ((WidevineOemProvisioning*)in)->oemDevicePrivateKey      ||
+               0 == ((WidevineOemProvisioning*)in)->oemDevicePrivateKeyLength   ||
+               NULL == ((WidevineOemProvisioning*)in)->oemDeviceCertificate     ||
+               0 == ((WidevineOemProvisioning*)in)->oemDeviceCertificateLength) {
+               ERROR("null private key/certor or size is  0");
+               return SA_STATUS_NULL_PARAMETER;
+            }
+
+            wvProvision = calloc(sizeof(WidevineOemProvisioning), 1);
+            if(NULL == wvProvision) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+
+            /*deep copy*/
+            wvProvision->oemDevicePrivateKeyLength =
+               ((WidevineOemProvisioning*)in)->oemDevicePrivateKeyLength;
+            wvProvision->oemDeviceCertificateLength =
+               ((WidevineOemProvisioning*)in)->oemDeviceCertificateLength;
+
+            wvProvision->oemDevicePrivateKey =
+               (void*)calloc(wvProvision->oemDevicePrivateKeyLength, 1);
+            if(NULL == wvProvision->oemDevicePrivateKey) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+            memcpy(wvProvision->oemDevicePrivateKey,
+                   ((WidevineOemProvisioning*)in)->oemDevicePrivateKey,
+                   wvProvision->oemDevicePrivateKeyLength);
+
+            /*here just provide an examlpe to pass certification data,
+             how to use certification data is all up to SOC vendors.
+            */
+            wvProvision->oemDeviceCertificate =
+               (void*)calloc(wvProvision->oemDeviceCertificateLength, 1);
+            if(NULL == wvProvision->oemDeviceCertificate) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+            memcpy(wvProvision->oemDeviceCertificate,
+                   ((WidevineOemProvisioning*)in)->oemDeviceCertificate,
+                   wvProvision->oemDeviceCertificateLength);
+
+            INFO("wvProvision:%x",wvProvision);
+            INFO("keyLen : %d", wvProvision->oemDevicePrivateKeyLength);
+            INFO("certLen : %d", wvProvision->oemDeviceCertificateLength);
+
+            if(0 == (memcmp(wvProvision->oemDevicePrivateKey, "SIGN", 4))) {
+               wvProvision->oemDevicePrivateKey += 8;
+               wvProvision->oemDevicePrivateKeyLength -=8;
+            } else {
+              INFO("NO SIGN in the header");
+            }
+
+            sa_key *rsa_key = create_uninitialized_sa_key();
+            assert(NULL != rsa_key);
+            status = sa_key_import(rsa_key, SA_KEY_FORMAT_EXPORTED,
+                        wvProvision->oemDevicePrivateKey,
+                        wvProvision->oemDevicePrivateKeyLength,NULL);
```

Comment:
Shouldn't the parameters field provide the SA_KEY_FORMAT_SOC type to denote the key is protected with the SOC key?